mono/metadata/native-library-qcall.c

```diff
@@ -0,0 +1,97 @@
+#include "config.h"
+#include "mono/metadata/assembly-internals.h"
+#include "mono/metadata/class-internals.h"
+#include "mono/metadata/icall-decl.h"
+#include "mono/metadata/loader-internals.h"
+#include "mono/metadata/loader.h"
+#include "mono/metadata/object-internals.h"
+#include "mono/metadata/reflection-internals.h"
+#include "mono/utils/checked-build.h"
+#include "mono/utils/mono-compiler.h"
+#include "mono/utils/mono-logger-internals.h"
+#include "mono/utils/mono-path.h"
+#include "mono/metadata/native-library.h"
+#include "mono/metadata/qcalllist.h"
+
+
+const int c_nECClasses = sizeof(c_qcalls)/sizeof(c_qcalls[0]);
+
+static gboolean is_end_of_array (MonoQCallFunc *func)  { return !!((int)func->flags & func_flag_end_of_array); }
+static gboolean has_signature (MonoQCallFunc *func)  { return !!((int)func->flags & func_flag_has_signature); }
+static gboolean is_unreferenced (MonoQCallFunc *func){ return !!((int)func->flags & func_flag_unreferenced); }
+static gboolean is_qcall (MonoQCallFunc *func)       { return !!((int)func->flags & func_flag_qcall); }
+//CorInfoIntrinsics   IntrinsicID(ECFunc *func)   { return (CorInfoIntrinsics)((INT8)(func->m_dwFlags >> 16)); }
+//int                 DynamicID(ECFunc *func)     { return (int)              ((int8)(func->m_dwFlags >> 24)); }
+
+static MonoQCallFunc* next_in_array(MonoQCallFunc *func)
+{
+    return (MonoQCallFunc*)((char*)func +sizeof(MonoQCallFunc));
+        //(HasSignature(func) ? sizeof(ECFunc) : offsetof(ECFunc, func->m_pMethodSig)));
+}
+
+static int 
+find_impls_index_for_class (MonoMethod* method)
+{
+    const char* namespace = m_class_get_name_space(method->klass);
+    const char* name = m_class_get_name(method->klass);
+
+    if (name == NULL)
+        return -1;
+
+    unsigned low  = 0;
+    unsigned high = c_nECClasses;
+
+#ifdef DEBUG
+    static bool checkedSort = false;
+    if (!checkedSort) {
+        checkedSort = true;
+        for (unsigned i = 1; i < high; i++)  {
+            int cmp = strcmp(c_qcalls[i].class_name, c_qcalls[i-1].class_name);
+            if (cmp == 0)
+                cmp = strcmp(c_qcalls[i].namespace_name, c_qcalls[i-1].namespace_name);
+            g_assert(cmp > 0);
+        }
+    }
+#endif // DEBUG
+    while (high > low) {
```

Comment:
Please figure out how to use `qsort` or `std::sort` here, instead of inlining it.
If you need inlining for perf, take an existing `qsort` and change it to a slightly ugly but fair enough macro. It's been done, it isn't hard, and it can be argued for perf.